### PR TITLE
Support for remote path in reindex api

### LINKED ISSUE #22913: Support for remote path in reindex api
**Elasticsearch version**: 5.x

**Plugins installed**: []

**JVM version**: 

**OS version**:

**Describe the feature**:
In the 5.x reindex api, support for remote is added, e.g, 
POST _reindex
{
  "source": {
    "remote": {
      "host": "http://otherhost:9200",
      "username": "user",
      "password": "pass"
    },
...

However the host property is strictly enforced to three parts, scheme, host and port. We host elasticsearch behind a proxy and need the ability to specify a path too. I found code in https://github.com/elastic/elasticsearch/blob/v5.2.0/modules/reindex/src/main/java/org/elasticsearch/index/reindex/RestReindexAction.java, that enforces this rule.

...
 private static final Pattern HOST_PATTERN = Pattern.compile("(?<scheme>[^:]+)://(?<host>[^:]+):(?<port>\\d+)");
...

Can we add the support for path?

### PR DESCRIPTION
**Pull request by vladimirdolzhenko**
Support for remote path in reindex api 
Closes #22913

### CODE REVIEW

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/RestReindexAction.java [hunk 1]
```diff
@@ -57,7 +57,7 @@
  */
 public class RestReindexAction extends AbstractBaseReindexRestHandler<ReindexRequest, ReindexAction> {
     static final ObjectParser<ReindexRequest, Void> PARSER = new ObjectParser<>("reindex");
-    private static final Pattern HOST_PATTERN = Pattern.compile("(?<scheme>[^:]+)://(?<host>[^:]+):(?<port>\\d+)");
+    private static final Pattern HOST_PATTERN = Pattern.compile("(?<scheme>[^:]+)://(?<host>[^:]+):(?<port>\\d+)(?<pathPrefix>(/.*)?)");
```
Comment (nik9000) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/RestReindexAction.java]: Is `(?<pathPrefix>/.*)?` good enough?

File: server/src/main/java/org/elasticsearch/index/reindex/RemoteInfo.java [hunk 2]
```diff
@@ -62,10 +63,16 @@
     private final TimeValue connectTimeout;
 
     public RemoteInfo(String scheme, String host, int port, BytesReference query, String username, String password,
-            Map<String, String> headers, TimeValue socketTimeout, TimeValue connectTimeout) {
+                      Map<String, String> headers, TimeValue socketTimeout, TimeValue connectTimeout) {
+        this(scheme, host, port, null, query, username, password, headers, socketTimeout, connectTimeout);
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/index/reindex/RemoteInfo.java]: Generally I'd prefer not to add method overrides for this unless we have a zillion uses of the old method. Would you be ok with fixing all of the call sites? Maybe in a followup PR?

**Review by nik9000 — COMMENTED** [review]
I left two minor things but otherwise it looks good to me.

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/RestReindexAction.java (hunk 1 above)
Comment (vladimirdolzhenko) [reply modules/reindex/src/main/java/org/elasticsearch/index/reindex/RestReindexAction.java]: @nik9000 no, `pathPrefix` could be really empty, like `http://somehost:9092` is a valid url - no extra `/` there

File: server/src/main/java/org/elasticsearch/index/reindex/RemoteInfo.java (hunk 2 above)
Comment (vladimirdolzhenko) [reply server/src/main/java/org/elasticsearch/index/reindex/RemoteInfo.java]: I totally agree with you - there are 11 are call sites - that stopped me to add extra parameter to any of that call site + 10 args... maybe worth to introduce RemoteInfo.Builder for that purpose ?

**Comment by vladimirdolzhenko** [comment]
test this please

**Review by nik9000 — APPROVED** [review]
LGTM.

Thanks for fixing this!

**Comment by vladimirdolzhenko** [comment]
Thank @nik9000 for the review

**Pull request merged**